include/hc_printf.hpp

```diff
@@ -98,14 +88,14 @@ enum PrintfPacketDataType {
 class PrintfPacket {
 public:
   void clear()             [[hc,cpu]] { type = PRINTF_UNUSED; }
-  void set(uint32_t d)     [[hc,cpu]] { type = PRINTF_UINT32_T;       data.ui = d; }
-  void set(int32_t d)      [[hc,cpu]] { type = PRINTF_INT32_T;        data.i = d; }
+  void set(uint32_t d)     [[hc,cpu]] { type = PRINTF_UINT32_T;       data.uli = d; }
+  void set(int32_t d)      [[hc,cpu]] { type = PRINTF_INT32_T;        data.li = d; }
   void set(uint64_t d)     [[hc,cpu]] { type = PRINTF_UINT64_T;       data.uli = d; }
   void set(int64_t d)      [[hc,cpu]] { type = PRINTF_INT64_T;        data.li = d; }
   void set(unsigned long long d) [[hc,cpu]] { type = PRINTF_UINT64_T; data.uli = d; }
   void set(long long d)    [[hc,cpu]] { type = PRINTF_INT64_T;        data.li = d; }
-  void set(hc::half d)     [[hc,cpu]] { type = PRINTF_HALF;           data.h = d; }
-  void set(float d)        [[hc,cpu]] { type = PRINTF_FLOAT;          data.f = d; }
+  void set(hc::half d)     [[hc,cpu]] { type = PRINTF_HALF;           data.d = d; }
+  void set(float d)        [[hc,cpu]] { type = PRINTF_FLOAT;          data.d = d; }
```

Comment:
Casting these to double would add a conversion which changes their value and their bit pattern.  If for example the format specifier is to print in hex that this will print out the wrong value.

Reply:
There are a number of things to consider here:

1. The current printf is a template, but the actual printf is supposed to be a variadic function. When calling a variadic function, "standard promotions" are applied automatically, where all floating point types smaller than double are converted to double. This is what we will see when we eventually have the new printf.
2. The current printf converts data.h to float anyway before printing via std::cout.
3. There is no format specifier for printing the hex of a half value in CUDA. If we introduce one, then it is the job of this specifier to convert the argument to half before printing. Usual considerations about rounding mode apply in that case.